[prometheus-pushgateway] Use a map for the extraArgs value in the pushgateway chart, handle alternate web prefix

devec0 opened the pull request:
This PR allows the use of an alternate web prefix, using the same mechanism as the standalone prometheus chart to conditionally modify the liveness and readiness checks on the pushgateway deployment if an alternate web root is configured.

This is required as otherwise, the checks fail with a 404 error using the current hard-coded check paths, as these paths must change if the web prefix is customised.

Signed-off-by: James Hebden <james@ec0.io>

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

monotek reviewed (changes requested):
Imho web.route-prefix should not be a part of extraArgs type should not be changed with a bugfix version as this would be a breaking change.

devec0 commented:
> Imho web.route-prefix should not be a part of extraArgs and extraArgs type should not be changed with a bugfix version as this would be a breaking change.

Thank you for looking at these changes @monotek! 

I've fixed the version number to better represent that project's versioning.

I also agree with your thought that the usage of the extraArgs map for controlling the prefix is awkward. I implemented it this way however to ensure consistency between this and the pushegateway configuration in the standalone prometheus chart in this repository. In my opinion, it would be cleaner to simply have a `webPrefix:` value (potentially under a `pushgateway` top level key) which was checked conditionally in the `deployment.yaml` and `servicemonitor.yaml` templates and its value used - in the `deployment.yaml`, the value could be checked and then the appropriate argument injected into the args list for the container, then any extraArgs (in list form, rather than map) could be appended if specified.

I'm happy to go either way, if you're prefer I implement it in the way I've described above, I can absolutely do that.

monotek commented:
> > Imho web.route-prefix should not be a part of extraArgs and extraArgs type should not be changed with a bugfix version as this would be a breaking change.
> 
> Thank you for looking at these changes @monotek!
> 
> I've fixed the version number to better represent the project's versioning.
> 
> I also agree with your thought that the usage of the extraArgs map for controlling the prefix is awkward. I implemented it this way however to ensure consistency between this and the pushegateway configuration in the standalone prometheus chart in this repository. In my opinion, it would be cleaner to simply have a `webPrefix:` value (potentially under a `pushgateway` top level key) which was checked conditionally in the `deployment.yaml` and `servicemonitor.yaml` templates and its value used - in the `deployment.yaml`, the value could be checked and then the appropriate argument injected into the args list for the container, then any extraArgs (in list form, rather than map) could be appended if specified.
> 
> I'm happy to go either way, if you're prefer I implement it in the way I've described above, I can absolutely do that.

Sounds good :)
